enhancements/insights/on-demand-data-gathering.md

```diff
@@ -0,0 +1,255 @@
+// data gatherer.
+type GathererStatus struct {
+	// conditions provide details on the status of each gatherer.
+	// +listType=atomic
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
+	Conditions []metav1.Condition `json:"conditions"`
+	// name is the name of the gatherer.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MaxLength=256
+	// +kubebuilder:validation:MinLength=5
+	Name string `json:"name"`
+	// lastGatherDuration represents the time spent gathering.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Type=string
+	// +kubebuilder:validation:Pattern="^([1-9][0-9]*(\\.[0-9]+)?(ns|us|µs|ms|s|m|h))+$"
+	LastGatherDuration metav1.Duration `json:"lastGatherDuration"`
+}
+
+```
+
+### Workflow Description
+
+1. User/cluster admin creates a new CR instance of the `datagather.insights.openshift.io` CRD. 
+2. The Insights operator is informed (using informers) about the new CR instance and spawns a new data gathering in a new container. 
+3. The new container process gathers Insights data based on the configuration provided in the CR instance and is responsible for uploading the gathered Insights archive and downloading the latest Insights analysis result from the console.redhat.com. 
+4. The new container process update the status attributes in the corresponding CR instance and stops.
+
+This scenario brings few questions and things to be discussed. Please see the following sections [Implementation details](#implementation-detailsnotesconstraints), [Risk mitigations](#risks-and-mitigations) and [Open questions](#open-questions)
+
+### API Extensions
+
+This is adding new Insights CRD in the `insights.openshift.io/v1alpha1` . The proposal of the structure of the new CRD is mentioned in the [Proposal](#Proposal) section.
+
+### Implementation Details/Notes/Constraints
+
+The Insights operator currently runs and gathers the data in a single container. 
+Uploading of new archives, pruning old archives as well as downloading the last Insights analysis report (this is always per archive) is implemented in this one container/binary. 
+When there is a successful upload (to the console.redhat.com) of the archive then there is some short delay (waiting for the processing in console.redhat.com pipelines) and then there is request to get the last Insights analysis report per given cluster. 
+Finally the status section of the `insightsoperator.operator.openshift.io` CR is updated. Note that there is a retry logic in case of failed upload and after some time (and number of attempts) the Insights operator is marked as degraded.
+Introducing a new on demand process/container for running an "extra" gather process raises the question about parallelism. It should be feasible to allow two (or more) parallel gather processes, but it is very important to keep in mind that only one (and that is the latest) Insights analysis report will be available.  
```

Comment:
This is a limitation of console.redhat.com? Who runs the service there? Is there any way to RFE that you could look up the analysis based on an ID of some description? To allow checking history/parallel analysis?

If not, you may want to have `DataGathers` sit in a `Pending` state until the existing `Running` `DataGather` has downloaded it's analysis, then run them sequentially, always processing based on creationTimestamp in a FIFO manner

Reply:
It was designed this way. The processing pipeline is maintained and developed by the CCX team. So theoretically yes we could raise an RFE, but the more important question is the use case (and the why). I can't see any benefit or reason to have Insights history available in a cluster. 

Yes allowing only one gather process at a time is the way IMO. I still think it would be better to interrupt any gather process, because why we should queue them when only the latest result will be available? I am afraid the queue would allow even more easier way to overwhelm the console.redhat.com part (what if something enqueues tens or hundreds of requests).

Reply:
> I can't see any benefit or reason to have Insights history available in a cluster.

I think the fact that we are allowing customers to customise which gatherers they can enable/disable gives warrant to wanting to be able to have multiple versions at once. Eg, the insights need to be gathered for separate support cases, if those are being worked separately, you could end up with a fight as they cancel each others requests to get their own gather output